### PR TITLE
1360 encounter status UI

### LINKED ISSUE #1360: Encounter status changes
Encounter statues are not really handled right now and need to be updated in the data model as well as in the UI.

### Data model

- [x] Update the possible encounter statuses to:

- `Pending`
- `Canceled`
- `Visited`

- [x] An encounter always has to have a status set and the default status is `Pending`

From this statuses we can derive a "dynamic" status value as following:
- [ ] if the status is `Pending` and the encounter start datetime is in the past the dynamic display status is `Missed`
- [ ] if the status is `Pending` and the encounter start datetime is in the future (or close to `now`) the dynamic display status is `Scheduled`
- [x] Canceled just mean Canceled

### Encounter UI

A user must be able to change the encounter status in the UI

- [ ] add a dropdown in the header to update the status in the header

Use the same `Missed` and `Schedule` logic to change the status to pending depending on if the encounter is in the past or in the future:

- [ ] if the encounter is in the past and has status `Canceled` or `Visited` the user can select `Missed` to internally change the status to `Pending`
- [ ] if the encounter is in the future and has status `Canceled` or `Visited` the user can select `Scheduled` to change the status to `Pending`
- [ ] Prevent users to change status to `Visited` when the encounter is in the future (?)

- [ ] If the Encounter is in the past show a warning show to the user (?)

### Encounter Report Arguments
- [x] update report arguments to work with the new statues (might need a specialized control)

### PR DESCRIPTION
**Pull request by clemens-msupply**
Fixes #1360

# 👩🏻‍💻 What does this PR do? 
- Changes the encounter status be `Pending | Visited | Canceled`
- Adds a dropdown in the encounter header to change the encounter status
- Adds a StatusCrumbs control in the encounter footer to display the current status
- Better align the encounter header items, IMHO still not very good though, the datetime row could be positioned nicer...

Known issues:
- StatusCrumbs log shows "Order History" but can't currently be configured

# 🧪 How has/should this change been tested? 

UI should update when changing the encounter status:
- Save button becomes active when there where no previous changes
- StatusCrumbs should update

### CODE REVIEW

**Comment by CarlosNZ** [comment]
I'm getting errors when I try to create a new encounter due to the changed enum. I've run the schema update script, and `yarn generate` locally.

Can you perhaps merge `feature/programs` back into here to see if that helps?

**Comment by clemens-msupply** [comment]
> I'm getting errors when I try to create a new encounter due to the changed enum. I've run the schema update script, and `yarn generate` locally.
> 
> Can you perhaps merge `feature/programs` back into here to see if that helps?

could you please try this script? `png/dev_server_schema_update/dev_server_schema_update.sh` This should setup a local sqlite DB for the remote_server with the correct schema. I just tried it and it seems to work.

**Comment by CarlosNZ** [comment]
> could you please try this script? `png/dev_server_schema_update/dev_server_schema_update.sh` This should setup a local sqlite DB for the remote_server with the correct schema. I just tried it and it seems to work. So run the script and then start the remote server.

Oh, I was using an outdated programschema repo. Looks good now. :)

File: client/packages/system/src/Encounter/utils.ts [hunk 1]
```diff
@@ -3,30 +3,32 @@ import { LocaleKey, TypedTFunction, useTranslation } from '@common/intl';
...
-      return t('label.encounter-status-missed');
+    case EncounterNodeStatus.Visited:
+      return t('label.encounter-status-visited');
     default:
-      ((_: never) => {
-        // exhaustive check
-        _;
-      })(status);
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      return ((_: never) => '')(status);
```
Comment (CarlosNZ) [inline client/packages/system/src/Encounter/utils.ts:19]: Weird that this lint comment is required. The linter rules already has `'no-unused-vars': 0,`, very strange 🤔

Comment (CarlosNZ) [reply client/packages/system/src/Encounter/utils.ts:19]: Okay, I figured it out with Mark -- there are two eslint rules that are in conflict:
```
    'no-unused-vars': 'off',
    '@typescript-eslint/no-unused-vars': 'error',
```

Just remove one of them and make sure the remaining one is set to "off" or `0`.

You can actually use a more specific setting  -- [argsIgnorePattern](https://eslint.org/docs/latest/rules/no-unused-vars#argsignorepattern) to just ignore parameters starting with an underscore and still warn about the rest, which might be worth doing.

Comment (CarlosNZ) [reply client/packages/system/src/Encounter/utils.ts:19]: Yes, have tested and this works:
```
 'no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
```

Just as long as you remove the typescript-specific rule as well.

File: client/packages/system/src/Encounter/utils.ts [hunk 2]
```diff
@@ -3,30 +3,32 @@ import { LocaleKey, TypedTFunction, useTranslation } from '@common/intl';
...
+export const effectiveStatus = (
+  encounter: EncounterRowFragment,
+  t: TypedTFunction<LocaleKey>
+): string => {
+  const status = encounter.status;
+  if (!status) {
+    return '';
   }
-  return '';
+  return encounterStatusTranslation(status, t);
```
Comment (CarlosNZ) [inline client/packages/system/src/Encounter/utils.ts:31]: Not quite sure why we need this as a separate function. Can't you just roll it into the existing `encounterStatusTranslation` and just have empty string returned as the default?

File: client/packages/system/src/Patient/Encounter/CreateEncounterModal.tsx [hunk 3]
```diff
@@ -87,12 +87,7 @@ export const CreateEncounterModal: FC = () => {
       ...draft,
       startDatetime,
       createdBy: { id: user?.id ?? '', username: user?.name ?? '' },
-      status:
-        date && DateUtils.isFuture(date)
-          ? EncounterNodeStatus.Scheduled
-          : date && DateUtils.isPast(date)
-          ? EncounterNodeStatus.Missed
-          : draft?.status,
+      status: EncounterNodeStatus.Pending,
```
Comment (CarlosNZ) [inline client/packages/system/src/Patient/Encounter/CreateEncounterModal.tsx:90]: So there's no different default if setting a new encounter in the past now?

**Review by CarlosNZ — COMMENTED** [review]
Looks good, couple of small suggestions.

Also:
> Better align the encounter header items, IMHO still not very good though, the datetime row could be positioned nicer...

We should probably use CSSGrid to keep the Toolbar rows and columns neatly aligned. But that can wait...

File: client/packages/system/src/Encounter/utils.ts (hunk 1 above)
Comment (clemens-msupply) [reply client/packages/system/src/Encounter/utils.ts:19]: thanks! done that

File: client/packages/system/src/Encounter/utils.ts (hunk 2 above)
Comment (clemens-msupply) [reply client/packages/system/src/Encounter/utils.ts:31]: Thanks, there was some more code in there at some point... I simplified it.

File: client/packages/system/src/Patient/Encounter/CreateEncounterModal.tsx (hunk 3 above)
Comment (clemens-msupply) [reply client/packages/system/src/Patient/Encounter/CreateEncounterModal.tsx:90]: No, the statuses Missed and Scheduled must  be calculated on the fly now and an encounter can be marked accordingly (in a future PR). The reason is that you would have to somehow update the encounter statues automatically, e.g. in a background process. Moreover, one idea of the patient program is that the clinician is in control of all document changes but an automated status setter would need to change the encounter document without user interaction.

**Comment by clemens-msupply** [comment]
> Looks good, couple of small suggestions.
> 
> Also:
> 
> > Better align the encounter header items, IMHO still not very good though, the datetime row could be positioned nicer...
> 
> We should probably use CSSGrid to keep the Toolbar rows and columns neatly aligned. But that can wait...

Agree! think a grid would work best, even though the date time row is a bit special, e.g. start and end time would probably be in one cell spanning two columns

**Review by CarlosNZ — APPROVED** [review]
LGTM! Thanks

**Pull request merged**